llama : fix pre-tokenization of non-special added tokens

compilade opened the pull request:
MPT and OLMo use a NeoX-style tokenizer, which has pre-normalized spaces in its added_tokens, while Gemma uses encoded (non-normalized) spaces in its added_tokens, and also has some HTML tags in there.

In the current state of `src/llama.cpp`, the tokenizer tests fails for MPT and OLMo when using `parse_special == false`.

HuggingFace's `tokenizers` pre-tokenizes the `added_tokens` before other pre-tokenization happens, see <https://github.com/huggingface/tokenizers/blob/fdd26ba9a3f0c133427aab0423888cbde91362d7/tokenizers/src/tokenizer/mod.rs#L726>

I've changed `tokenizer_st_partition` to be called even if `parse_special` is false to pre-tokenize user-defined tokens, but it processes control tokens only when `parse_special` is true. **This allows pre-tokenizing non-special added tokens correctly at all times, while still allowing to safely not tokenize control tokens when `parse_special` is false.**

This also fixes Gemma's tokenization of HTML tags, but it requires re-conversion because the added tokens were previously not using the correct types, and also because user-defined tokens are assumed to use bare spaces by `llama.cpp`, but Gemma uses "▁" (normal tokens are not affected).

# TODO

- Tests should pass
  - [x] `./build/bin/test-tokenizer-0 ./models/ggml-vocab-mpt.gguf`
    - Note that NFC normalization is still a problem to be solved, but everything else works.
  - [x] `./build/bin/test-tokenizer-0 ./models/ggml-vocab-olmo.gguf`
  - [x] using `tests/test-tokenizer-random.py` with Gemma-2's tokenizer, there's no difference whatsoever to the reference `tokenizer.model`

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Medium

compilade committed:
llama : fix mpt and olmo pre-tokenizer

compilade commented:
> Can you confirm that the commit [68220fe](https://github.com/ggerganov/llama.cpp/commit/68220feaf8b8c8f8b9d1690fd975bba445d5d4b6) still fails with MPT and OLMO?

@jaime-m-p With that commit, with the MPT tokenizer, I can see that the spaces *do* get handled correctly with `tests/test-tokenizer-0.cpp`, although it seems to instead fail with the apostrophe in ` I've been` (and others like ` I'll make`), splitting it in ` I` `'` `ve` ` been` instead of using the `'ve` token), which was not a problem here.

However, correcting the `regex_exprs` of `LLAMA_VOCAB_TYPE_PRE_MPT` makes it work correctly in all the cases:

```diff
diff --git a/llama.cpp b/llama.cpp
index ab8620ec..58b84839 100644
--- a/llama.cpp
+++ b/llama.cpp
@@ -13222,13 +13222,7 @@ struct llm_tokenizer_bpe {
                 };
                 break;
             case LLAMA_VOCAB_PRE_TYPE_MPT:
-                // TODO: MPT pre-tokenization regexes are unknown
-                //       the following are close, but not exact. run the following:
-                //       ./bin/test-tokenizer-0 ../models/ggml-vocab-mpt.gguf
-                GGML_ASSERT("MPT pre-tokenization regexes are unknown - fixes needed");
                 regex_exprs = {
-                    "\\s?\\p{L}+",
-                    "\\s?\\p{P}+",
                     "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)",
                 };
                 break;
```

What you've done is definitely better than this PR, since this means MPT and OLMo and other NeoX-style tokenizers can simply directly use the GPT-2 pre-tokenizer, like they should.

I think I'll close this in favor of #8039.

compilade commented:
From further testing, it seems like passing `parse_special=true` to `llama_tokenize` makes the correct multi-space tokens be used for NeoX-style tokenizers. That's what fixed it in <https://github.com/ggerganov/llama.cpp/commit/68220feaf8b8c8f8b9d1690fd975bba445d5d4b6>. ***However***, when *not* using `parse_special`, this doesn't work anymore and instead uses the other multi-space tokens.

This PR makes it still pick the special spaces even when `parse_special` is false. Not sure if that's desired, maybe not. It depends on how models based on that kind of tokenizer behave with the non-special multi-space tokens.

Seems like NeoX tokenizers complicate the decision of whether user input should be exempt of special tokens or not.

compilade commented on src/llama.cpp:
```diff
@@ -13985,6 +14002,10 @@ struct llm_tokenizer_bpe {
     void tokenize(const std::string & text, std::vector<llama_vocab::id> & output) {
         int final_prev_index = -1;
 
+        // FIXME: pre-tokenize added_tokens (user-defined tokens) before other pre-tokenization
```
Hmm, yes! This is it. But it makes me think there should be a distinction between special tokens and user-defined tokens. I think `parse_special=false` should *still* pre-tokenize non-special user-defined tokens before the rest.

compilade committed:
Merge branch 'master' into compilade/fix-mpt-pretok

compilade committed:
llama : pre-tokenize non-special user-defined tokens first

compilade committed:
Merge branch 'master' into compilade/fix-mpt-pretok

compilade committed:
llama : fix detection of control-like user-defined tokens

compilade committed:
convert_hf : identify which user-defined tokens are control tokens

Only used in _set_vocab_gpt2() for now.

compilade committed:
convert_hf : identify more added control tokens for SPM tokenziers

This makes Gemma and Gemma-2 tokenize pretty much EVERYTHING correctly,
including HTML tags and consecutive spaces,
but it unfortunately requires model re-conversion.

There seems to be a weird behavior of the HF tokenizer for Gemma,
which prefers to use the 16-space token over more lengthy space tokens,
while using the SentencePiece tokenizer does not do this.
(the implementation in llama.cpp has the same behavior as SentencePiece)

* llama : fix wrong pre-tokenization of byte tokens

compilade commented:
In trying to fix the pre-tokenization for non-special added tokens for MPT and OLMo, I've managed to also fix Gemma and Gemma-2's pre-tokenization.

Gemma on `master` doesn't pre-tokenize HTML tags correctly, while in this branch it does! Unfortunately, it requires re-conversion because the user-defined tokens in the rest of the codebase are assumed to use bare spaces instead of `▁`.

There's a weird behavior of the HF tokenizer for Gemma which seems to prefer the 16-space token (id: 153) over longer ones:
```
INFO:test-tokenizer-random:generator_custom_text: ini
ERROR:test-tokenizer-random: Expected: [2, 153, 153, 155, 132133, 14130] ['<bos>', '                ', '                ', '                  ', 'fifty', ' spaces']
ERROR:test-tokenizer-random:   Result: [2, 168, 156, 132133, 14130] ['<bos>', '                               ', '                   ', 'fifty', ' spaces']
```
But when using `tokenizer.model` instead of `tokenizer.json` for the reference tokenizer, this error does not happen.

So I assume what `llama.cpp` is now doing is the right thing.

bartowski1182 commented on convert_hf_to_gguf.py:
```diff
@@ -373,6 +373,29 @@ def from_model_architecture(cls, arch: str) -> type[Model]:
+        seems_special = seems_special or (token_text.startswith("<｜") and token_text.endswith("｜>"))  # deepseek-coder
+
+        # TODO: should these be marked as UNUSED instead? (maybe not)
+        seems_special = seems_special or (token_text.startswith("<unused") and token_text.endswith(">"))  # gemma{,-2}
```
should things like this be defined in the conversion script under the specific model to avoid accidental false hits? if, for some weird reason, a model comes around with a non-special token that starts with <|, would be annoying to avoid that

maybe does_token_look_special should take in 2 lists: 1 list of strings of known special tokens, and a list of tuples of starts/ends with tokens

So for gemma2, we'd call it with:

```
special_tokens = ["<mask>", "<2mass>", "[@BOS@]"]
special_tags = [("<unused", "|>")]

self.does_token_look_special(token, special_tokens, special_tags)
```

and then here we'd have:
```
seems_special = token_text in special_tokens
for start_tag, end_tag in special_tags:
  seems_special = seems_special or (token_text.startswith(start_tag) and token_text.endswith(end_tag))

return seems_special
```

bartowski1182 replied:
oh i realize I misread the structure of the code, hmmm.. still not impossible but would have to be passed at a higher level

compilade replied:
> if, for some weird reason, a model comes around with a non-special token that starts with <|, would be annoying to avoid that

This only affects added tokens either `added_tokens` in `tokenizer.json` or `added_tokens_decoder` from `tokenizer_config.json`, so it **should not affect normal tokens** starting with `<|` in any way. And added tokens starting with `<|` and ending with `|>` are arguably always control tokens; this was added pretty much *because* some model makers wrongly marked those as non-special (notably, `<|User|>`, `<|Assistant|>` and `<|EOT|>` in `deepseek-coder` are supposedly non-special).

I did not yet notice any conflict in the `added_tokens` to justify making model-specific checks instead of always checking for all known "special-but-arent-marked-special" tokens.

compilade committed:
llama : fix Viking pre-tokenizer regex

The order was previously wrong, which caused errors in some tests.

compilade committed:
llama : fix command-r detokenization

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -657,6 +669,25 @@ def _create_vocab_sentencepiece(self):
+                for token_id, token_data in added_tokens_decoder.items():
+                    token_id = int(token_id)
+                    token: str = token_data["content"]
+                    if toktypes[token_id] != SentencePieceTokenTypes.UNKNOWN:
```
`toktypes` is initialized with UNKNOWN, so this still works.

https://github.com/ggerganov/llama.cpp/blob/d6fe269ced93d45783a3b37c3cb20554264e5578/convert_hf_to_gguf.py#L637

compilade commented on src/llama.cpp:
```diff
@@ -5640,7 +5640,7 @@ static void llm_load_vocab(
     // build special tokens cache
     {
         for (llama_vocab::id id = 0; id < (llama_vocab::id)n_vocab; ++id) {
-            if (!(vocab.id_to_token[id].attr & LLAMA_TOKEN_ATTR_NORMAL)) {
+            if (vocab.id_to_token[id].attr & (LLAMA_TOKEN_ATTR_CONTROL | LLAMA_TOKEN_ATTR_USER_DEFINED)) {
```
Should UNKNOWN tokens only be parsed when `parse_special == true`, or all the time like `USER_DEFINED` tokens?

I'm not sure if UNKNOWN tokens should be specially parsed at all.

I would tend toward only parsing them with `parse_special == true`, like CONTROL tokens. (this is also the current behavior of `master`)

I'm trying to figure out where UNKNOWN tokens are used and if it's useful to specially parse them. But this might differ from HF's `tokenizers`, so I need a model with UNKNOWN tokens to test this out. If I don't find one, I can always modify one.

compilade replied:
> but we need parse UNKNOWN token.

I've fixed this in <https://github.com/ggerganov/llama.cpp/commit/98edea60bcd67458d70138c13c428414cd8ec63a>. UNKNOWN tokens are parsed when `parse_special == true`, as on `master`.

compilade committed:
convert_hf : reduce usages of the UNKNOWN token type

compilade committed:
llama : add UNKNOWN tokens in the special tokens cache

compilade commented:
> Which models are going to need re-conversion? Just gemma2? deepseek2?

@oldmanjk

Yes, Gemma, Gemma-2, Command-R, Command-R-Plus, and deepseek-coder (not sure which of their models use that pre-tokenizer) need re-conversion, because this changes the type of some of their `USER_DEFINED` tokens to `CONTROL` for a safer tokenization when `parse_special == false` (e.g. with `deepseek-coder`, `<|EOT|>` was previously a `USER_DEFINED` token while it should have been marked as a `CONTROL` token). This is only relevant if you're tokenizing with `parse_special == false`, but most of the ways you're likely using the models use `parse_special == true`.

It's mostly Gemma and Gemma-2 which ***really*** need re-conversion for correctness (unfortunately), because all non-special `added_tokens` were previously marked as `NORMAL` in the GGUF models, while they should have been `USER_DEFINED`, so that pre-tokenization works correctly. (this mostly affects HTML tag tokenization in the case of Gemma (and Gemma 2))

compilade commented:
> So deepseek only really needs re-conversion if you're "tokenizing with parse_special == false?" What does that mean? When might someone do that?

This only happens when using `--in-prefix` and/or `--in-suffix` with `llama-cli` (aka `main.cpp`), because the other ways chat templates are handled always use `parse_special == true`.  

`llama-server` always uses `parse_special == true` for now, so nothing really changes over there.

> deepseek-coder-v2

I checked, and <https://huggingface.co/deepseek-ai/DeepSeek-Coder-V2-Instruct> is a bit less affected by this than the older `deepseek-coder` models. At least this one sets `<|EOT|>` as special, but there are still some other special-looking `added_tokens` which are not marked special (like `<｜User｜>` and `<｜Assistant｜>`), but which should arguably have been.

> Each time takes my 13900K and 4090 about four days. The fans are driving my family insane and I'd love to finish this project before the heat death of the universe...

@oldmanjk It's not necessary to reconvert, especially if your use-case is not affected, and *especially* if it's *that* demanding. You're likely fine with your existing `deepseek-coder-v2` model.

Makes me think there should really be a way to update the metadata of GGUF models without rebuilding them all over again. (`gguf_set_metadata.py` has its limits, and `gguf_new_metadata.py` writes a copy of the model, but they both don't allow "applying" what `convert_hf_to_gguf.py` would do. This does seem like something to think about.)

compilade commented:
> to release these imatrices and/or quants, one should still probably re-convert. 

`imatrix` uses `parse_special == true`, so it's likely not affected either.

compilade commented:
> What exactly are the special tokens?

@oldmanjk It's the `added_tokens` in `tokenizer.json`. Some are `USER_DEFINED`, some are `CONTROL` tokens. But they're all treated differently than `NORMAL` tokens. To get a list of them, assuming you have `jq`, and `grep` on a Unix-like system:

```console
$ jq .added_tokens < tokenizer.json
```

If you want to `grep` for special tokens which look like control tokens in a particular text file:
```console
$ jq --raw-output0 .added_tokens[].content < ./path/to/tokenizer.json | grep -z '^<' | xargs -I{} -0 grep --color -F {} ./path/to/dataset.txt
```
While if you want to search for *any* special token, simply remove the middle `grep -z '^<'` like so:

```console
$ jq --raw-output0 .added_tokens[].content < ./path/to/tokenizer.json | xargs -I{} -0 grep --color -F {} ./path/to/dataset.txt
```

compilade committed:
Merge branch 'master' into compilade/fix-mpt-pretok

compilade committed:
convert_hf : reduce usages of UNKNOWN for InternLM2

This makes the changes from #8321 more consistent
with the other changes made here.

oldgithubman reviewed (changes requested):
I don't really know what I'm doing, so forgive me if this isn't helpful

compilade commented on convert_hf_to_gguf.py:
```diff
@@ -373,6 +373,29 @@ def from_model_architecture(cls, arch: str) -> type[Model]:
         except KeyError:
             raise NotImplementedError(f'Architecture {arch!r} not supported!') from None
 
+    def does_token_look_special(self, token: str | bytes) -> bool:
```
I prefer not to make it a `@staticmethod`, to allow overriding it in the subclasses of `Model` if needed.

compilade commented on src/llama.cpp:
```diff
@@ -16185,12 +16176,20 @@ struct fragment_buffer_variant {
 // #define PRETOKENIZERDEBUG
 
-static void tokenizer_st_partition(const llama_vocab & vocab, std::forward_list<fragment_buffer_variant> & buffer) {
+static void tokenizer_st_partition(const llama_vocab & vocab, std::forward_list<fragment_buffer_variant> & buffer, bool parse_special) {
```
> "function 'tokenizer_st_partition' declared 'static', move to anonymous namespace instead"

I've seen this warning for pretty much every `static` function in `src/llama.cpp`. Fixing this is out of the scope of this PR.

compilade committed:
test-tokenizer-random : reduce potential confilcts with #8379

* test-tokenizer-random : add a failing edge case for falcon

compilade commented on tests/test-tokenizer-random.py:
```diff
@@ -458,8 +458,8 @@ def check_detokenizer(text: str, text1: str, text2: str) -> bool:
-            logger.error(" Expected: " + str(ids1))
-            logger.error("   Result: " + str(ids2))
+            logger.error(" Expected: " + str(ids1) + f" {[tokenizer1.decode([id]) for id in ids1]}")
+            logger.error("   Result: " + str(ids2) + f" {[tokenizer2.decode([id]) for id in ids2]}")
```
I've reverted these two lines in <https://github.com/ggerganov/llama.cpp/pull/8228/commits/59ce85318aeace3d100395b86061fd10b3af6f54>, even though they were useful for debugging, because this is a test script anyway, and because these lines also conflicts with #8379

compilade commented:
> Instead of modifying the vocab words, there is an alternative (doing it the other way around).
Just pre-normalize the input so added tokens with spaces can be found in the raw added tokens:

I agree it would be better to keep the vocab text verbatim. But unconditionally pre-escaping the input would lose the distinction between `"\x20"` and `"\xe2\x96\x81"`. I think this can only be fixed with token-level attributes like `LLAMA_TOKEN_ATTR_NORMALIZED` once they'll be included in GGUF models, and then conditionally normalizing the tokens when using the vocab. Otherwise this would break tokenization for `deepseek-coder`.

> ```c
>             case LLAMA_VOCAB_TYPE_WPM:
>             case LLAMA_VOCAB_TYPE_SPM:
>             case LLAMA_VOCAB_TYPE_UGM: {
>                 std::string unscaped_text = llama_unescape_whitespace(token_text);
>                 return _try_copy(unscaped_text.data(), unscaped_text.size());
>             }
> ```

For control tokens, I'm not sure if the whitespaces should be unescaped.
For example, `deepseek-coder` uses `<｜begin▁of▁sentence｜>`, with the `"\xe2\x96\x81"` spaces. I just ran a test with that token in a string, as well as `<｜fim▁begin｜>`, and the escaped spaces are kept with HF `tokenizers` (same as current `llama.cpp` behavior). The same text without the escaped spaces does not match these tokens. (These tokens are marked as `"normalized": true` in the added_tokens)

I also don't see a way around special-casing BYTE tokens with `llama_token_to_byte`. These tokens are usually in the form `<0x[0-9A-Fa-f]{2}>`

I think it should be more like:

```c
switch (llama_vocab_get_type(model->vocab)) {
    case LLAMA_VOCAB_TYPE_WPM:
    case LLAMA_VOCAB_TYPE_SPM:
    case LLAMA_VOCAB_TYPE_UGM: {
        if (attr & LLAMA_TOKEN_ATTR_NORMALIZED) {
            return _try_copy(token_text.data(), token_text.size());
        } else if (attr & LLAMA_TOKEN_ATTR_BYTE) {
            char byte = (char) llama_token_to_byte(model->vocab, token);
            return _try_copy((char*) &byte, 1);
        } else {
            std::string result = token_text;
            llama_unescape_whitespace(result);
            return _try_copy(result.data(), result.size());
        }
        break;
    }
```
